sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -793,6 +817,56 @@ case class Aggregate(
     copy(child = newChild)
 }
 
+object Aggregate {
+  private def collectComplexGroupingExpressions(groupingExpressions: Seq[Expression]) = {
+    groupingExpressions.zipWithIndex
+      .foldLeft(mutable.Map.empty[Expression, (Expression, Int)]) {
```

Comment:
It looks weird to use `foldLeft` with a mutable map. How about just creates the mutable map at the beginning and write a for loop? Then it's more readable to Java people.